Append cucumber report pt2

BryceStevenWilley opened the pull request:
A second attempt of #540, to fix #491.

Adds a JS script that concatenates the printed out cucumber report to the end of the Kiln report, stripping out any terminal specific color commands.

Feel free to critique my JS and give improvements, this was mostly from following the tutorials on [Node.js](https://nodejs.dev/learn).

BryceStevenWilley committed:
Quick and dirty JS script (47 lines) that does the same thing and the 3 line bash script

BryceStevenWilley committed:
Was using unreleased features of cucumber! lol

Everything before this PR is good for us: https://github.com/cucumber/cucumber-js/pull/1794/files

BryceStevenWilley committed:
Keeping the action in sync

BryceStevenWilley committed:
clean_report.js referening cucumber.js as well

plocket commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "test": "npm run unit && npm run langs_setup && npm run cucumber -- $@",
-    "cucumber_base": "./node_modules/.bin/cucumber-js --require ./lib/index.js",
+    "cucumber_base": "run_cucumber(){ ./node_modules/.bin/cucumber-js \"$@\"; ./lib/utils/clean_reports.js; }; run_cucumber",
```
I'm not completely sure how scope works in bash. Can we give this function another name? It's also ok if we change the name of the other function instead. For that reason and because it's strange to me that they're both "running cucumber". They're not doing the same thing, so what are they each doing?

plocket replied:
Your code makes me wonder if we really need the `--` in the other function. Would appreciate your thoughts on that.

BryceStevenWilley replied:
I think it's less about bash and more about how npm is actually running it. I assume it just copy pastes everything in the string directly to shell, which does overwrite it with the latest line. A quick example:
 ```bash
$ run_hi() { echo 'hi 1'; }
$ run_hi
hi 1
$ run_hi() { echo 'hi 2'; }
$ run_hi
hi 2
 ```
 Not opposed to changing the name though, I just needed to wrap everything in some function to make sure `$@` was stable. (might not be necessary, but idk for sure).`
 
 > Your code makes me wonder if we really need the -- in the other function
 
TBH, I don't know what exactly `--` does in this case, it's very program dependent, and I can't find what it does for cucumber at all.

plocket reviewed (commented):
The tests are passing and I'm not completely sure how to test their functionality. I'm not seeing the cucumber part of the report inside the report file on GitHub.

I have some very bad guesses. For example, the async nature of cucumber.js (which we might call cucumber-report.js?). Not sure the node command line is waiting for it. That could have worked locally because it should have finished executing by the time a human person checks the local file, but might not work on GitHub, which closes everything up. I believe it's possible to do all those operations synchronously, so that might be something to try, though I admit I prefer knowing the source of a problem for sure.

I'm also trying to read up on `readdirSync` to see if it can use relative paths. I've mostly seen people use `__dirname`. I suspect that's not the case or you would have noticed it breaking locally.

If you already know the answers to some of these, then it'd be helpful in saving some time. This also may be a more productive discussion in real time.

BryceStevenWilley commented:
I do think it's working! The last lines of the report that I can see are:
> 54 scenarios (54 passed)
> 313 steps (313 passed)
> 5m15.634s (executing steps: 5m15.349s)

Which is what is printed by the cucumber summary process when nothing fails. Compare that with [the latest merge on releases/v4](https://github.com/SuffolkLITLab/ALKiln/actions/runs/2080517700), which doesn't have that summary of scenarios/steps passed and the time it took everything.

BryceStevenWilley committed:
Renamed run_cucumber to run_base

plocket commented:
Ah, my report file wasn't loading all the way somehow in finder! Looks great!

plocket commented:
We want it in the log, though, right? Not the report?

BryceStevenWilley committed:
Add errors to log, not to kiln report

Simplifies the clean report script, since there will only ever be
one debug log.

BryceStevenWilley commented:
Apologies, that fact got a bit lost in trying to figure all of the mess, but it's a quick fix. Waiting for the tests to confirm we're adding to the `debug_log.txt` now.

mcdonaldcarolyn committed:
pin versions of dependencies in action.yml and package.json (#548)

plocket commented:
debug_log.txt is confirmed 👍

plocket committed:
Update vulnerabilities, bump for pin versions and debug_log.txt

BryceStevenWilley committed:
Added description of the appending to Changelog

plocket committed:
Merge branch 'append_cucumber_report_pt2' of github.com:SuffolkLITLab/ALKiln into append_cucumber_report_pt2

plocket commented:
Newest wrinkle: I ran a deliberately failing tests. The tests failed correctly, but the run passed for some reason: https://github.com/SuffolkLITLab/ALKiln/runs/5835449414?check_suite_focus=true

BryceStevenWilley committed:
return the original exit code from cucumber

Not the one from clean_reports.js, which should always succeed

BryceStevenWilley commented:
Confirmed that manually making the workflow fail does make the tests fail! https://github.com/SuffolkLITLab/ALKiln/actions/runs/2098401294

plocket commented on .gitignore:
```diff
@@ -129,3 +131,4 @@ _al_test_project_name*
 package-lock.json
 #report logs
 /debug_log.txt
+/log.txt
```
Sorry, at this point log stuff needs to be removed. I didn't catch anything elsewhere, but this can be cleaned up.

```suggestion
/debug_log.txt
```

BryceStevenWilley replied:
This was intentional, I had rebased and accudentally commited my log.txt to the repo. I think that others will likely make the same mistake, and IMO it's doesn't need to be removed from gitignore anyway.

plocket replied:
I think there are only three of us with the problem and we've all dealt with it, but it won't do any harm. Sound good then 👍

plocket reviewed (approved):
So close and ever closer! Just a little cleanup and then this is a great addition to output! It's a non-breaking cleanup, so I'll approve and you can remove or not and merge as you see fit.
